_posts/2024-07-19-crowdstrike-is-the-opposite-of-cheriot.markdown

```diff
@@ -0,0 +1,48 @@
+---
+layout: post
+title:  "Crowdstrike is the opposite of CHERIoT"
+date:   2024-07-03
+categories: security philosophy
+author: "David Chisnall"
+---
+
+The news today is full of reports about a Windows kernel driver from CrowdStrike causing large numbers of computers to be unbootable.
+This entire problem is a comedy of errors that stems from a development model that is the exact opposite of the philosophy behind CHERIoT
+
+On a Windows system, the absolute highest privilege mode is the virtual secure mode, which is used for a fairly limited set of core system services.
+Below that is kernel mode, which is almost as powerful.
+A bug in kernel mode can access any file on the system, tamper with any user data, and (as today's news showed) cause the entire system to fail.
+
+Secure software development starts with the principle of least privilege.
+No software should be in the kernel unless it absolutely needs to be.
+The first mistake for the CrowdStrike bug was putting code in the kernel that didn't need to be there.
+In CHERIoT, the equivalent code is the switcher, which handles transitions between compartments and between threads.
+This is a very small amount of code, [from a single file](https://github.com/microsoft/cheriot-rtos/blob/main/sdk/core/switcher/entry.S).
+
+It appears, from reverse engineering reports, that the problem arose from CrowdStrike parsing a corrupted rules file and dereferencing an invalid pointer.
+In other words, it's the kind of memory-safety bug that CHERI was designed to protect against.
+CHERI does not prevent you from doing this kind of thing (though language features can make it much harder).
+On a CHERI system such as CHERIoT, this would still trap.
+The difference is the impact.
```

Comment:
I might try reworking this to be more of a Socratic kind of thing.  "But wait, you might ask, CHERI _traps_ on invalid memory accesses; isn't that exactly what's happening here?"